src/Umbraco.Core/Persistence/NPocoSqlExtensions.cs

```diff
@@ -1,143 +1,225 @@
+            var tableNameAttribute = type.FirstAttribute<TableNameAttribute>();
+            string tableName = tableNameAttribute == null ? string.Empty : tableNameAttribute.Value;
+
+            return sql.LeftOuterJoin(sqlSyntax.GetQuotedTableName(tableName));
+        }
+
+        public static Sql.SqlJoinClause RightJoin<T>(this Sql sql, ISqlSyntaxProvider sqlSyntax)
+        {
+            var type = typeof(T);
+            var tableNameAttribute = type.FirstAttribute<TableNameAttribute>();
+            string tableName = tableNameAttribute == null ? string.Empty : tableNameAttribute.Value;
+
+            return sql.RightJoin(sqlSyntax.GetQuotedTableName(tableName));
+        }
+
+        public static Sql On<TLeft, TRight>(this Sql.SqlJoinClause sql, ISqlSyntaxProvider sqlSyntax, Expression<Func<TLeft, object>> leftMember,
+                                           Expression<Func<TRight, object>> rightMember, params object[] args)
+        {
+            var leftType = typeof(TLeft);
+            var rightType = typeof(TRight);
+            var leftTableName = leftType.FirstAttribute<TableNameAttribute>().Value;
+            var rightTableName = rightType.FirstAttribute<TableNameAttribute>().Value;
+
+            var left = ExpressionHelper.FindProperty(leftMember) as PropertyInfo;
+            var right = ExpressionHelper.FindProperty(rightMember) as PropertyInfo;
+            var leftColumnName = left.FirstAttribute<ColumnAttribute>().Name;
+            var rightColumnName = right.FirstAttribute<ColumnAttribute>().Name;
+
+            string onClause = string.Format("{0}.{1} = {2}.{3}",
+                sqlSyntax.GetQuotedTableName(leftTableName),
+                sqlSyntax.GetQuotedColumnName(leftColumnName),
+                sqlSyntax.GetQuotedTableName(rightTableName),
+                sqlSyntax.GetQuotedColumnName(rightColumnName));
+            return sql.On(onClause);
+        }
+
+        public static Sql OrderByDescending(this Sql sql, params object[] columns)
+        {
+            return sql.Append(new Sql("ORDER BY " + String.Join(", ", (from x in columns select x + " DESC").ToArray())));
+        }
+
+        public static Sql.SqlJoinClause LeftOuterJoin(this Sql sql, string table) { return Join(sql, "LEFT OUTER JOIN ", table); }
+        public static Sql.SqlJoinClause RightJoin(this Sql sql, string table) { return Join(sql, "RIGHT JOIN ", table); }
+
+        // copied from NPoco (private)
+        private static Sql.SqlJoinClause Join(Sql sql, string joinType, string table)
+        {
+            return new Sql.SqlJoinClause(sql.Append(new Sql(joinType + table)));
+        }
+
+        public static Sql SelectCount(this Sql sql)
+        {
+            return sql.Select("COUNT(*)");
+        }
+
+        public static Sql Select<T>(this Sql sql, Func<RefSql, RefSql> refexpr = null)
+        {
+            Database database = ApplicationContext.Current.DatabaseContext.Database; // fixme.npoco
```

Comment:
We cannot use singletons here if we need the db object it needs to be passed to the method.